### PR TITLE
fix bug #19

### PR DESCRIPTION
**Pull request by lourkeur**
It seems to fix the issue and not break tests.  There's still 128kB of leaked memory reported by `binding/lua/leak_test.lua` on my machine, but it seems to happen both before and after this commit and I'm not up to speed with the memory leak issue so I don't know how to interpret that.  I also didn't touch `Session.L` since it's unused.

Feedback appreciated

### CODE REVIEW

File: binding/lua/lmpack.c [hunk 1]
```diff
@@ -826,6 +828,7 @@ static int lmpack_session_receive(lua_State *L)
   if (session->unpacker != LUA_REFNIL) {
     lmpack_geti(L, session->reg, session->unpacker);
     unpacker = lmpack_check_unpacker(L, -1);
+  unpacker->L = L;
```
Comment (justinmk) [inline binding/lua/lmpack.c]: indent

Comment (lourkeur) [inline binding/lua/lmpack.c]: nice catch. fixed.

**Comment by tarruda** [comment]
Thanks for the fix. Can you add a test that reproduces the bug describe in #19?


> I'm not up to speed with the memory leak issue so I don't know how to interpret that. I also didn't touch Session.L since it's unused.

I assume you mean the `leak_check.lua` script. What version of lua you are using to run it? Is it reproducible in lua 5.1.5?

**Comment by lourkeur** [comment]
>Thanks for the fix. Can you add a test that reproduces the bug describe in #19?
My pleasure. Here are 3 simple regression tests.

>I assume you mean the leak_check.lua script. What version of lua you are using to run it? Is it reproducible in lua 5.1.5?
Mainly 5.3.3 on my machine (I have three versions available), and the leak doesn't appear to happen on 5.1.5. Were you aware of such a leak? When writing my previous message I believed the leak was already being discussed, but I can't find where I got that from. Sorry for being confusing.
So:
+ It seems this PR by itself doesn't break [the travis lua job](https://travis-ci.org/tarruda/libmpack/jobs/191986843). (I had to make a small temporary fix 38d6451fa4a591f0bfee1196ce0e958744956c39)
+ memory leak goes -> #22

**Comment by tarruda** [comment]
@lourkeur I'd like to make a couple of changes before merging. Can you give me permission to edit the PR branch?(See https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/)

**Comment by lourkeur** [comment]
@tarruda No problem. I can't login until saturday though (nor push), so unless it can be done over email we'll have to wait. I'm on my smartphone.


-------- Message d'origine --------
De : Thiago de Arruda <notifications@github.com> 
Date : 15/01/2017  21:24  (GMT+01:00) 
À : tarruda/libmpack <libmpack@noreply.github.com> 
Cc : lourkeur <louis@bettens.org>, Mention <mention@noreply.github.com> 
Objet : Re: [tarruda/libmpack] fix bug #19 (#21) 

@lourkeur I'd like to make a couple of changes before merging. Can you give me permission to edit the PR branch?(See https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/)

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.


  
  




{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/tarruda/libmpack","title":"tarruda/libmpack","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/tarruda/libmpack"}},"updates":{"snippets":[{"icon":"PERSON","message":"@tarruda in #21: @lourkeur I'd like to make a couple of changes before merging. Can you give me permission to edit the PR branch?(See https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/)"}],"action":{"name":"View Pull Request","url":"https://github.com/tarruda/libmpack/pull/21#issuecomment-272721084"}}}

**Comment by tarruda** [comment]
> @tarruda No problem. I can't login until saturday though (nor push), so unless it can be done over email we'll have to wait. I'm on my smartphone.

Don't worry, there's no rush.

But you don't need to push, all you need to do is give me permission to edit the PR's branch

**Comment by lourkeur** [comment]
@tarruda ok, I just ticked the checkbox.

**Comment by tarruda** [comment]
Hmm, since you used the master branch to send the pull request, something must have gone wrong with my git command to update the PR's branch and it ended up pushing my master to your master(which caused the issue to be closed since there are no commits).

Can you open a new PR based on the current master? Here's the diff from the original PR:

```diff
diff --git a/binding/lua/lmpack.c b/binding/lua/lmpack.c
index 5749465..3388612 100644
--- a/binding/lua/lmpack.c
+++ b/binding/lua/lmpack.c
@@ -433,6 +433,7 @@ static int lmpack_unpacker_unpack(lua_State *L)
     return luaL_error(L, "expecting between 2 and 3 arguments");

   unpacker = lmpack_check_unpacker(L, 1);
+  unpacker->L = L;

   str_init = str = luaL_checklstring(L, 2, &len);
   startpos = lua_gettop(L) == 3 ? luaL_checknumber(L, 3) : 1;
@@ -720,6 +721,7 @@ static int lmpack_packer_pack(lua_State *L)
     return luaL_error(L, "expecting exactly 2 arguments");

   packer = lmpack_check_packer(L, 1);
+  packer->L = L;
   packer->root = lmpack_ref(L, packer->reg);
   luaL_buffinit(L, &buffer);
   b = luaL_prepbuffer(&buffer);
@@ -831,6 +833,7 @@ static int lmpack_session_receive(lua_State *L)
   if (session->unpacker != LUA_REFNIL) {
     lmpack_geti(L, session->reg, session->unpacker);
     unpacker = lmpack_check_unpacker(L, -1);
+    unpacker->L = L;
     rcount += 2;
     lua_pop(L, 1);
   }
diff --git a/binding/lua/test.lua b/binding/lua/test.lua
index 9146ba8..2f22c82 100644
--- a/binding/lua/test.lua
+++ b/binding/lua/test.lua
@@ -96,6 +96,14 @@ describe('mpack', function()
         assert.are_same(object, unpacked)
       end)
     end)
+
+    describe('when used across coroutine boundaries', function()
+      it('has the same behavior as when not', function()
+        unpacked, pos = coroutine.wrap(function() return unpack(msgpack) end)()
+        assert.are_same(object, unpacked)
+        assert.are_same(#msgpack + 1, pos)
+      end)
+    end)
   end)

   describe('pack', function()
@@ -117,6 +125,15 @@ describe('mpack', function()
         assert.are_same(object, o)
       end)
     end)
+
+    describe('when used across coroutine boundaries', function()
+      it('has the same behavior as when not', function()
+        packed = coroutine.wrap(function() return pack(object) end)()
+        local unpack = mpack.Unpacker()
+        local o = unpack(packed)
+        assert.are_same(object, o)
+      end)
+    end)
   end)

   describe('handling cyclic references', function()
@@ -367,6 +384,15 @@ describe('mpack', function()
         {session:receive(fhex('61'))})
     end)

+    describe('when used across coroutine boundaries', function()
+      it('has the same behavior as when not', function()
+        local input = fhex('94 00 00 a6 6d 65 74 68 6f 64 92 01 a1 61')
+        local received = {'request', 0, 'method', {1, 'a'}, 15}
+        local session = mpack.Session({unpack = mpack.Unpacker()})
+        assert.are_same(received, coroutine.wrap(function() return {session:receive(input)} end)())
+      end)
+    end)
+
     it('grows session capacity', function()
       local session = mpack.Session()
       assert.are_same(fhex('94 00 00'), session:request())
```

**Comment by tarruda** [comment]
(BTW the reason I didn't fix it myself is because my permission to update the PR's branch is gone after the issue is closed)

**Pull request closed without merging**